File: app/schemas/milestone.py
from pydantic import BaseModel, ConfigDict, Field, field_validator
from datetime import date
from typing import Optional, List
from app.core.constants import NOTE_MAX_LENGTH
from app.utils.s3 import sign_image_urls
from app.schemas.achievement import UnlockedAchievementInfo

class MilestoneBase(BaseModel):
    milestone_type: str = Field(..., max_length=50)
    title: str = Field(..., max_length=100)
    achieved_date: date
    image_urls: List[str] = []
    notes: Optional[str] = Field(None, max_length=NOTE_MAX_LENGTH)

class MilestoneCreate(MilestoneBase):
    pass

class MilestoneUpdate(BaseModel):
    milestone_type: Optional[str] = Field(None, max_length=50)
    title: Optional[str] = Field(None, max_length=100)
    achieved_date: Optional[date] = None
    image_urls: Optional[List[str]] = None
    notes: Optional[str] = Field(None, max_length=NOTE_MAX_LENGTH)

class MilestoneResponse(MilestoneBase):
    id: int
    baby_id: int
    user_id: Optional[int]
    unlocked_achievements: List[UnlockedAchievementInfo] = []

    @field_validator("image_urls", mode="after")
    @classmethod
    def secure_image_urls(cls, v: List[str]) -> List[str]:
        return sign_image_urls(v)

    model_config = ConfigDict(from_attributes=True)
class MilestoneTimelineGroup(BaseModel):
    month_age: int
    milestones: List[MilestoneResponse]

File: frontend/schemas/milestone.ts
import { z } from "zod"

export const milestoneSchema = z.object({
  milestone_type: z.string().min(1, "達成項目を選択してください"),
  title: z.string().min(1, "マイルストーン名を入力してください"),
  achieved_date: z.string().min(1, "達成日を選択してください"),
  image_urls: z.array(z.string()).max(10, "写真は最大10枚までです"),
  notes: z.string().optional().nullable(),
})

export type MilestoneFormValues = z.infer<typeof milestoneSchema>
